src/imageio/format/jxl.c

```diff
@@ -0,0 +1,491 @@
+  dt_lua_register_module_member(darktable.lua_state.state, self, dt_imageio_jxl_params_t, lossless, gboolean);
+
+  dt_lua_register_module_member(darktable.lua_state.state, self, dt_imageio_jxl_params_t, quality, float);
+
+  dt_lua_register_module_member(darktable.lua_state.state, self, dt_imageio_jxl_params_t, encoding_effort, int);
+
+  dt_lua_register_module_member(darktable.lua_state.state, self, dt_imageio_jxl_params_t, decoding_effort, int);
+#endif
+}
+
+void cleanup(dt_imageio_module_format_t *self)
+{
+}
+
+
+
+const char *mime(dt_imageio_module_data_t *data)
+{
+  return "image/jxl";
+}
+
+const char *extension(dt_imageio_module_data_t *data)
+{
+  return "jxl";
+}
+
+int dimension(struct dt_imageio_module_format_t *self, struct dt_imageio_module_data_t *data, uint32_t *width,
+              uint32_t *height)
+{
+  *width = 1073741823;
+  *height = 1073741823;
+  return 1;
+}
+
+int bpp(dt_imageio_module_data_t *data)
+{
+  return 32;
+}
+
+int write_image(struct dt_imageio_module_data_t *data, const char *filename, const void *in_tmp,
+                dt_colorspaces_color_profile_type_t over_type, const char *over_filename, void *exif, int exif_len,
+                int imgid, int num, int total, struct dt_dev_pixelpipe_t *pipe, const gboolean export_masks)
+{
+  // By default error code
+  int out = 1;
+
+  const float *in = (float *)in_tmp;
+  float *pixels = NULL;
+  uint8_t *out_buf = NULL;
+  FILE *out_file = NULL;
+
+#define JXL_ASSERT(code)                                                                                          \
+  {                                                                                                               \
+    JxlEncoderStatus res = code;                                                                                  \
+    if(res != JXL_ENC_SUCCESS)                                                                                    \
+    {                                                                                                             \
+      fprintf(stderr, "JXL call failed with err %d (src/imageio/format/jxl.c#L%d)\n", res, __LINE__);             \
+      goto end;                                                                                                   \
+    }                                                                                                             \
+  }
+
+  dt_imageio_jxl_params_t *params = (dt_imageio_jxl_params_t *)data;
+  int width = params->global.width;
+  int height = params->global.height;
+
+  JxlEncoder *encoder = JxlEncoderCreate(NULL);
+
+  // JxlThreadParallelRunnerDefaultNumWorkerThreads()
+  uint32_t num_threads = JxlResizableParallelRunnerSuggestThreads(width, height);
+  void *runner = JxlThreadParallelRunnerCreate(NULL, num_threads);
+  JXL_ASSERT(JxlEncoderSetParallelRunner(encoder, JxlThreadParallelRunner, runner));
+
+  cmsHPROFILE out_profile = dt_colorspaces_get_output_profile(imgid, over_type, over_filename)->profile;
```

Comment:
Perhaps we should also consider setting the color encoding the "native" way using JxlEncoderSetColorEncoding() if possible first, see also `avif.c`?

Reply:
What would be the benefit of doing this? I would just have to manually gather all the data that is already included in the icc profile and provide that to libjxl instead right?

Reply:
There might be a risk that some 3rd party apps don't handle ICC profiles for example, so the "native" way of encoding the color profile in the stream might be preferable? This seems to be the case w/ AVIF/HEIF images, but then again, that might be the legacy of the video side of those containers where video players of course don't usually understand ICC profiles so there was always a preference for the "nclx" way of encoding the color space (this is btw using the same CICP enums as JPEG XL). OTOH, some 3rd party apps behave the opposite (ICC is renederd fine, while CICP is not)...

See e.g these for potential pitfalls:

https://discuss.pixls.us/t/experimenting-with-hlg-and-hdr10-for-still-images/23361/17
https://discuss.pixls.us/t/experimenting-with-hlg-and-hdr10-for-still-images/23361/19

I guess there should be a way for the JXL output plugin to generate both, just like the current AVIF one does based on the output profile selection.

Reply:
Thank You, that makes more sense.

Could you also clarify the purpose of the `if(imgid > 0)` check just above the code you sent (https://github.com/darktable-org/darktable/blob/master/src/imageio/format/avif.c#L284).
I saw the same check in the code of lots of the other output format modules but I didn't understand the significance and didn't do anything similar in the jxl module.

Reply:
> Could you also clarify the purpose of the `if(imgid > 0)` check just above

Not sure, sorry; we'll have to wait for someone else to chip in on that one...

Reply:
The icc profile can be big making the image bigger. The nclx container just takes a few bytes. We want to safe bandwidth with those new formats!

Reply:
jxl does not use the "nclx" box in a container, but encodes all non-ICC, non-Exif etc. _rendering_ related metadata in the codestream directly if you chose not go with the other option, so you actually have to decode it first even inside exiv2 just to read that metadata :( I'd keep the ICC option supported, a few kB is not that big deal for high res HDR images (jxl plans to brotli compress ICC and Exif boxes anyway)...